File: src-tauri/src/core/diagnostics.rs
use std::path::Path;

#[derive(Debug, Clone, Copy, PartialEq, Eq, serde::Serialize, serde::Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum DiagnosticSeverity {
    Info,
    Warn,
    Error,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, serde::Serialize, serde::Deserialize)]
pub enum DiagnosticArea {
    #[serde(rename = "import_sgdk")]
    ImportSgdk,
    #[serde(rename = "import_gamemaker")]
    ImportGameMaker,
    #[serde(rename = "import_mugen")]
    ImportMugen,
    #[serde(rename = "import_ikemen")]
    ImportIkemen,
    #[serde(rename = "import_openbor")]
    ImportOpenBor,
    #[serde(rename = "build_sgdk")]
    BuildSgdk,
    #[serde(rename = "build_snes")]
    BuildSnes,
    #[serde(rename = "libretro_emulation")]
    LibretroEmulation,
    #[serde(rename = "runtime_setup")]
    RuntimeSetup,
    #[serde(rename = "hardware")]
    Hardware,
    #[serde(rename = "project")]
    Project,
    #[serde(rename = "codegen")]
    Codegen,
}

#[derive(Debug, Clone, PartialEq, Eq, serde::Serialize, serde::Deserialize)]
pub struct ActionableDiagnostic {
    pub severity: DiagnosticSeverity,
    pub area: DiagnosticArea,
    pub source_path: Option<String>,
    pub line: Option<u32>,
    pub column: Option<u32>,
    pub user_message: String,
    pub technical_detail: String,
    pub suggested_action: String,
    pub blocking: bool,
    pub evidence_path: Option<String>,
}

impl ActionableDiagnostic {
    pub fn new(
        severity: DiagnosticSeverity,
        area: DiagnosticArea,
        user_message: impl Into<String>,
        technical_detail: impl Into<String>,
        suggested_action: impl Into<String>,
        blocking: bool,
    ) -> Self {
        Self {
            severity,
            area,
            source_path: None,
            line: None,
            column: None,
            user_message: user_message.into(),
            technical_detail: technical_detail.into(),
            suggested_action: suggested_action.into(),
            blocking,
            evidence_path: None,
        }
    }

    pub fn blocking_error(
        area: DiagnosticArea,
        user_message: impl Into<String>,
        technical_detail: impl Into<String>,
        suggested_action: impl Into<String>,
    ) -> Self {
        Self::new(
            DiagnosticSeverity::Error,
            area,
            user_message,
            technical_detail,
            suggested_action,
            true,
        )
    }

    pub fn with_source_path(mut self, source_path: impl Into<String>) -> Self {
        self.source_path = Some(source_path.into());
        self
    }

    pub fn with_line_column(mut self, line: Option<u32>, column: Option<u32>) -> Self {
        self.line = line;
        self.column = column;
        self
    }

    pub fn with_evidence_path(mut self, evidence_path: impl Into<String>) -> Self {
        self.evidence_path = Some(evidence_path.into());
        self
    }
}

pub fn target_build_area(target: &str) -> DiagnosticArea {
    if target.eq_ignore_ascii_case("snes") {
        DiagnosticArea::BuildSnes
    } else {
        DiagnosticArea::BuildSgdk
    }
}

pub fn build_diagnostics_from_log(
    target: &str,
    log: &[crate::compiler::build_orch::BuildLogLine],
    evidence_path: Option<&Path>,
) -> Vec<ActionableDiagnostic> {
    let error_lines = log
        .iter()
        .filter(|line| line.level == "error")
        .map(|line| line.message.trim())
        .filter(|message| !message.is_empty())
        .collect::<Vec<_>>();

    if error_lines.is_empty() {
        return vec![build_failure_diagnostic(
            target,
            "Build falhou sem linhas de erro estruturadas no log.",
            evidence_path,
        )];
    }

    error_lines
        .into_iter()
        .map(|message| build_failure_diagnostic(target, message, evidence_path))
        .collect()
}

pub fn build_failure_diagnostic(
    target: &str,
    technical_detail: &str,
    evidence_path: Option<&Path>,
) -> ActionableDiagnostic {
    let area = target_build_area(target);
    let lower = technical_detail.to_ascii_lowercase();
    let source_path = extract_quoted_path(technical_detail);
    let (line, column) = parse_line_column(technical_detail);

    let (user_message, suggested_action) = if lower.contains("asset referenciado nao encontrado") {
        let asset = source_path
            .as_deref()
            .unwrap_or("asset referenciado pela cena");
        (
            format!("Build falhou porque o asset '{}' nao foi encontrado.", asset),
            "Restaure o arquivo ausente ou atualize a entidade para apontar para um asset existente.".to_string(),
        )
    } else if lower.contains("toolchain sgdk nao encontrada") {
        (
            "Build falhou porque a toolchain SGDK nao esta disponivel.".to_string(),
            "Abra Runtime Setup e instale SGDK oficial, ou configure SGDK_ROOT para a pasta correta.".to_string(),
        )
    } else if lower.contains("toolchain pvsneslib nao encontrada") {
        (
            "Build SNES falhou porque a toolchain PVSnesLib nao esta disponivel.".to_string(),
            "Abra Runtime Setup e instale PVSnesLib oficial, ou configure PVSNESLIB_HOME para a pasta correta.".to_string(),
        )
    } else if lower.contains("build abortado: erros de hardware constraints") {
        (
            "Build foi bloqueado por violacoes de hardware do target.".to_string(),
            "Abra os avisos de hardware, reduza os recursos marcados como fatais e rode a validacao novamente.".to_string(),
        )
    } else if lower.contains("falha ao gerar ast") || lower.contains("falha ao resolver prefabs") {
        (
            "Build falhou durante a preparacao do modelo canonico da cena.".to_string(),
            "Abra a cena, confira prefabs/grafos referenciados e salve novamente antes de recompilar.".to_string(),
        )
    } else {
        (
            "Build falhou durante a emissao, staging ou execucao da toolchain.".to_string(),
            "Veja o detalhe tecnico, corrija o arquivo ou dependencia indicada e rode Build & Run novamente.".to_string(),
        )
    };

    let mut diagnostic = ActionableDiagnostic::blocking_error(
        area,
        user_message,
        technical_detail,
        suggested_action,
    )
    .with_line_column(line, column);

    if let Some(source_path) = source_path {
        diagnostic = diagnostic.with_source_path(source_path);
    }
    if let Some(path) = evidence_path {
        diagnostic = diagnostic.with_evidence_path(path.to_string_lossy().to_string());
    }
    diagnostic
}

fn extract_quoted_path(message: &str) -> Option<String> {
    let start = message.find('\'')?;
    let remainder = &message[start + 1..];
    let end = remainder.find('\'')?;
    let candidate = remainder[..end].trim();
    (!candidate.is_empty()).then(|| candidate.to_string())
}

fn parse_line_column(message: &str) -> (Option<u32>, Option<u32>) {
    for token in message.split_whitespace() {
        let parts = token.split(':').collect::<Vec<_>>();
        if parts.len() < 3 {
            continue;
        }
        let line = parts
            .get(parts.len().saturating_sub(3))
            .and_then(|value| value.parse::<u32>().ok());
        let column = parts
            .get(parts.len().saturating_sub(2))
            .and_then(|value| value.parse::<u32>().ok());
        if line.is_some() {
            return (line, column);
        }
    }
    (None, None)
}

File: src/core/diagnostics.ts
import type { BuildResult } from "./ipc/buildService";

export type DiagnosticSeverity = "info" | "warn" | "error";

export type DiagnosticArea =
  | "import_sgdk"
  | "import_gamemaker"
  | "import_mugen"
  | "import_ikemen"
  | "import_openbor"
  | "build_sgdk"
  | "build_snes"
  | "libretro_emulation"
  | "runtime_setup"
  | "hardware"
  | "project"
  | "codegen";

export interface ActionableDiagnostic {
  severity: DiagnosticSeverity;
  area: DiagnosticArea;
  source_path: string | null;
  line: number | null;
  column: number | null;
  user_message: string;
  technical_detail: string;
  suggested_action: string;
  blocking: boolean;
  evidence_path: string | null;
}

export const DIAGNOSTIC_AREA_LABELS: Record<DiagnosticArea, string> = {
  import_sgdk: "Import SGDK",
  import_gamemaker: "Import GameMaker",
  import_mugen: "Import MUGEN",
  import_ikemen: "Import Ikemen",
  import_openbor: "Import OpenBOR",
  build_sgdk: "Build SGDK",
  build_snes: "Build SNES",
  libretro_emulation: "Libretro",
  runtime_setup: "Runtime Setup",
  hardware: "Hardware",
  project: "Projeto",
  codegen: "CodeGen",
};

export const DIAGNOSTIC_SEVERITY_LABELS: Record<DiagnosticSeverity, string> = {
  info: "Info",
  warn: "Warn",
  error: "Error",
};

const IMPORT_AREAS = new Set<DiagnosticArea>([
  "import_sgdk",
  "import_gamemaker",
  "import_mugen",
  "import_ikemen",
  "import_openbor",
]);

export function diagnosticConsoleMessage(diagnostic: ActionableDiagnostic): string {
  return `${diagnostic.user_message} Acao recomendada: ${diagnostic.suggested_action}`;
}

export function diagnosticCopyText(diagnostic: ActionableDiagnostic): string {
  const location = [
    diagnostic.source_path ? `Arquivo: ${diagnostic.source_path}` : null,
    diagnostic.line ? `Linha: ${diagnostic.line}` : null,
    diagnostic.column ? `Coluna: ${diagnostic.column}` : null,
    diagnostic.evidence_path ? `Evidencia: ${diagnostic.evidence_path}` : null,
  ]
    .filter(Boolean)
    .join("\n");

  return [
    `[${DIAGNOSTIC_AREA_LABELS[diagnostic.area]} / ${DIAGNOSTIC_SEVERITY_LABELS[diagnostic.severity]}]`,
    diagnostic.user_message,
    `Acao recomendada: ${diagnostic.suggested_action}`,
    location,
    "Detalhe tecnico:",
    diagnostic.technical_detail,
  ]
    .filter((part) => part && part.length > 0)
    .join("\n");
}

export function importAreaForProfile(profileId: string): DiagnosticArea {
  switch (profileId) {
    case "gamemaker":
      return "import_gamemaker";
    case "mugen":
      return "import_mugen";
    case "ikemen_go":
      return "import_ikemen";
    case "openbor":
      return "import_openbor";
    default:
      return "project";
  }
}

export function buildAreaForTarget(target: string): DiagnosticArea {
  return target === "snes" ? "build_snes" : "build_sgdk";
}

export function createFallbackDiagnostic(input: {
  area: DiagnosticArea;
  technicalDetail: string;
  sourcePath?: string | null;
  evidencePath?: string | null;
  severity?: DiagnosticSeverity;
  blocking?: boolean;
  suggestedAction?: string;
}): ActionableDiagnostic {
  const severity = input.severity ?? "error";
  const blocking = input.blocking ?? severity === "error";
  const technicalDetail = input.technicalDetail || "Erro nao detalhado.";
  const sourcePath = input.sourcePath ?? extractQuotedPath(technicalDetail);
  const { line, column } = parseLineColumn(technicalDetail);

  return {
    severity,
    area: input.area,
    source_path: sourcePath ?? null,
    line,
    column,
    user_message: fallbackUserMessage(input.area, sourcePath, technicalDetail),
    technical_detail: technicalDetail,
    suggested_action: input.suggestedAction ?? fallbackSuggestedAction(input.area),
    blocking,
    evidence_path: input.evidencePath ?? null,
  };
}

export function normalizeBuildDiagnostics(
  result: BuildResult,
  target: "megadrive" | "snes" | string,
  projectDir: string
): ActionableDiagnostic[] {
  if (result.diagnostics && result.diagnostics.length > 0) {
    return result.diagnostics;
  }

  const area = buildAreaForTarget(target);
  const errorLines = result.log
    .filter((line) => line.level === "error")
    .map((line) => line.message.trim())
    .filter(Boolean);

  if (errorLines.length === 0) {
    return [
      createFallbackDiagnostic({
        area,
        technicalDetail:
          "Build falhou sem linhas de erro estruturadas no log; verifique o historico completo do Console.",
        evidencePath: buildEvidencePath(projectDir, target),
      }),
    ];
  }

  return errorLines.map((technicalDetail) =>
    createFallbackDiagnostic({
      area,
      technicalDetail,
      sourcePath: extractQuotedPath(technicalDetail),
      evidencePath: buildEvidencePath(projectDir, target),
    })
  );
}

function fallbackUserMessage(
  area: DiagnosticArea,
  sourcePath: string | null | undefined,
  technicalDetail: string
): string {
  const lower = technicalDetail.toLowerCase();
  if (area === "build_sgdk" || area === "build_snes") {
    if (lower.includes("asset referenciado nao encontrado")) {
      return `Build falhou porque o asset ${sourcePath ?? "referenciado"} nao foi encontrado.`;
    }
    if (lower.includes("toolchain")) {
      return `Build falhou porque uma toolchain obrigatoria nao esta disponivel.`;
    }
    return "Build falhou durante a emissao, staging ou execucao da toolchain.";
  }

  if (area === "libretro_emulation") {
    return "Emulacao falhou porque a ROM ou o core Libretro nao pode ser carregado.";
  }

  if (area === "runtime_setup") {
    return "Runtime Setup falhou ao preparar uma dependencia oficial.";
  }

  if (IMPORT_AREAS.has(area)) {
    return `${importFailureLabel(area)} falhou porque a origem nao pode ser processada.`;
  }

  if (area === "hardware") {
    return "Build foi bloqueado por violacoes de hardware do target.";
  }

  return "Operacao falhou e precisa de acao antes de continuar.";
}

function importFailureLabel(area: DiagnosticArea): string {
  switch (area) {
    case "import_sgdk":
      return "Importacao SGDK";
    case "import_gamemaker":
      return "Importacao GameMaker";
    case "import_mugen":
      return "Importacao MUGEN";
    case "import_ikemen":
      return "Importacao Ikemen";
    case "import_openbor":
      return "Importacao OpenBOR";
    default:
      return "Importacao";
  }
}

function fallbackSuggestedAction(area: DiagnosticArea): string {
  if (area === "build_sgdk") {
    return "Abra o detalhe tecnico, corrija o asset/codigo indicado ou reinstale SGDK pelo Runtime Setup.";
  }
  if (area === "build_snes") {
    return "Abra o detalhe tecnico, corrija o asset/codigo indicado ou reinstale PVSnesLib pelo Runtime Setup.";
  }
  if (area === "runtime_setup") {
    return "Abra Runtime Setup, revise o item indicado e tente reinstalar a dependencia oficial.";
  }
  if (area === "libretro_emulation") {
    return "Verifique se a ROM e valida para o target e instale o core Libretro correspondente.";
  }
  if (IMPORT_AREAS.has(area)) {
    return "Verifique se a pasta doadora contem os arquivos raiz esperados e tente importar novamente.";
  }
  if (area === "hardware") {
    return "Reduza os recursos marcados como fatais e rode a validacao novamente.";
  }
  return "Revise o detalhe tecnico, corrija a causa indicada e tente novamente.";
}

function extractQuotedPath(message: string): string | null {
  const match = message.match(/'([^']+)'/);
  return match?.[1] ?? null;
}

function parseLineColumn(message: string): { line: number | null; column: number | null } {
  const match = message.match(/:(\d+):(\d+):/);
  return {
    line: match ? Number.parseInt(match[1], 10) : null,
    column: match ? Number.parseInt(match[2], 10) : null,
  };
}

function buildEvidencePath(projectDir: string, target: string): string | null {
  if (!projectDir) {
    return null;
  }
  return `${projectDir.replace(/[\\/]$/, "")}/build/${target === "snes" ? "snes" : "megadrive"}`;
}
